---
title: "Testing supercell_GE"
output: html_notebook
---

Test tree ways of averaging gene expression:
1) an old one with a loop 
2) a new one with an aggregate function 
3) a new one with sapply function 

Summary:
the slowest one is with aggregate function for sevaral reason, including that the gene expression must be a dataframe (not just a matrix), also, the it has to be transported twice. The second slowest is the old version using a loop and the fastest so far is a function using sapply after splitting membership vector into list of groups.

For this reason, supercell_ge(...) has been rewriten with a use of sapply function. The origianals of fucntions are stired in this fie before benchmarking


```{r}
library(microbenchmark)
library(Seurat)

N.g <- 10000
N.c.seq <- c(1000, 5000, 10000, 20000, 50000)
N.SC.seq <- c(500, 300, 100, 50)

set.seed(12345)
m.seurat <- readRDS("../../SuperCellExamples/data/Tcells_10x_Zheng/m_seurat.Rds")

keep.genes <- names(sort(rowSums(m.seurat@data>0), decreasing = T))[1:N.g]
ge <- m.seurat@data[keep.genes,]
rm(m.seurat)


```

I save here the original of used funcitons because after the analysis, function supercell_GE has been rewriten
```{r originals of function}
#### using loop
supercell_GE_loop  <- function(ge, groups, do.median.norm = FALSE){ # the old vertsion
  u.groups        <- unique(groups)
  u.groups        <- u.groups[!is.na(u.groups)]
  N.groups        <- length(u.groups)
  N.genes         <- nrow(ge)
  supercell.GE    <- matrix(0, nrow = N.genes, ncol = N.groups)
  rownames(supercell.GE) <- rownames(ge)
  colnames(supercell.GE) <- u.groups
  for(g in as.character(u.groups)){
    idxs             <- which(groups == g)
    #print(idxs)
    if(length(idxs) < 2){
      supercell.GE[,g] <- ge[,idxs]
    } else{
      if(N.genes > 1){
        supercell.GE[,g] <- apply(ge[,idxs], 1, mean)
      } else {
        supercell.GE[,g] <- mean(ge[,idxs])
      }
    }
  }
  if(!(TRUE %in% is.na(as.numeric(colnames(supercell.GE))))){
    supercell.GE <- supercell.GE[,order(as.numeric(colnames(supercell.GE)))]
  }

  if(do.median.norm){
    supercell.GE <- (supercell.GE+0.01)/apply(supercell.GE+0.01, 1, median)
  }
  return(supercell.GE)
}

#### using aggregate (requites a transported version of ge matrix, i.e. rows as cells and cols as genes)
supercell_GE_aggr  <- function(ge, groups, do.median.norm = FALSE){ # was expected to perform better that the old one, but tiil even longer to run
  u.groups        <- unique(groups)
  u.groups        <- u.groups[!is.na(u.groups)]
  N.groups        <- length(u.groups)
  N.genes         <- nrow(ge)

  supercell.GE             <- aggregate(ge, list(groups), mean)

  if(do.median.norm){
    supercell.GE <- (supercell.GE+0.01)/apply(supercell.GE+0.01, 1, median)
  }
  return(supercell.GE)
}


#### using split of membership (group) vector followed with sapply
supercell_GE_sapply  <- function(ge, groups, do.median.norm = FALSE){  #the fastest among three tested functions
  goups.idx <-  plyr:::split_indices(groups)

  fun <- function(idx){
      Matrix::rowMeans(ge[, idx, drop = FALSE])
  }
  supercell.GE             <- sapply(goups.idx, fun)

  if(do.median.norm){
    supercell.GE <- (supercell.GE+0.01)/apply(supercell.GE+0.01, 1, median)
  }
  return(supercell.GE)
}



```


```{r}
if(!exists("time.GE")) time.GE <- c()

for(N.c in N.c.seq){
  print(N.c)
  for(N.SC in N.SC.seq){
    print(N.SC)
    
    set.seed(12345)
    cur.membership <- c(1:N.SC, sample(N.SC, N.c - N.SC, replace = T))
    cur.membership.list <- plyr:::split_indices(cur.membership)

    print("Start loop")
    ge_loop <- microbenchmark("loop" = {loop <- supercell_GE_loop(ge[,1:N.c], groups = cur.membership)}, times = 1)
    row <- c((ge_loop$time/(10^9)), N.c, N.SC, "loop")
    time.GE <- rbind(time.GE, row)

    
    print("Start sapply")
    ge_sapply <-  microbenchmark("sapply" = {sap <- supercell_GE_sapply(ge[,1:N.c], groups = cur.membership)}, times = 1)
    row <- c((ge_sapply$time/(10^9)), N.c, N.SC, "sapply")
    time.GE <- rbind(time.GE, row)
  }
}

time.GE.df <- as.data.frame(time.GE)
colnames(time.GE.df) <- c("time_seconds", "N_cells", "N_supercells", "Method")

for(col in 1:3){
  time.GE.df[, col] <- as.numeric(as.vector(time.GE.df[, col]))
}

ggplot(time.GE.df, aes(x = N_supercells, y = time_seconds, alpha = factor(N_cells), color = Method)) + geom_point() + geom_line() +
  scale_color_manual(values = c("red", "blue"), labels = c("Using loop", "using sapply")) +
  scale_alpha_discrete(name = "N single cells") +
  scale_x_continuous(breaks = N.SC.seq) +
  labs(title = "Computational time of function supercell_GE") + 
  theme_bw() + theme(asp = 1)

ggsave("../plots/comp_time_supercell_GE.pdf")
```

